src/com/cdd/bao/validator/RemappingChecker.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * BioAssay Ontology Annotator Tools
+ *
+ * (c) 2014-2018 Collaborative Drug Discovery Inc.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License 2.0
+ * as published by the Free Software Foundation:
+ *
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+package com.cdd.bao.validator;
+
+import java.io.*;
+import java.util.*;
+
+/*
+	Use logic in this class to detect cycles in remapping directives. 
+*/
+
+public class RemappingChecker
+{
+	public static void validateRemappings(Map<String, String> remappings) throws IOException
+	{
+		// prohibit null-terminated chains
+		if (remappings.values().contains(null)) throw new IOException("Null-terminated remappings are not allowed.");
+
+		List<String> dependencies = new ArrayList<>();
+		for (String uri : remappings.keySet())
+		{
+			dependencies.clear();
+			while (remappings.containsKey(uri))
+			{
+				dependencies.add(uri); // add URI to dependency chain
+				if ((uri = remappings.get(uri)) == null) break; 
+				if (dependencies.contains(uri)) throw new IOException(buildMessage(dependencies));
```

Comment:
I think it's odd to throw an IOException. I know we convert all kinds of exceptions ultimately into IOExceptions, however it just feels wrong. This is more an `IllegalArgumentException`

Reply:
I'm inclined to agree, but borrowing IOException is less bad than making a top-level class that's only thrown once and never trapped explicitly. With a bit more thought put into the architecture, though, I might change my mind about that policy... e.g. if we made a "BrokenOntologyException" or something, used it in a bunch of places to communicate when the ontology source data is not working out, and trapping it to provide actionable feedback...